Fixing tests failures on non-defined facts

aerostitch opened the pull request:
Hi,

Currently the tests are failing on testing of undefined facts when compiling with STRICT_VARIABLES:
```
Undefined variable "::aio_agent_version"; Undefined variable "aio_agent_version" at /home/travis/build/voxpupuli/puppet-archive/spec/fixtures/modules/archive/manifests/params.pp:24 on node testing-worker-linux-docker-bc362139-3367-linux-14.prod.travis-ci.org
```
and:
```
Undefined variable "::archive_windir"; Undefined variable "archive_windir" at /home/travis/build/voxpupuli/puppet-archive/spec/fixtures/modules/archive/manifests/params.pp:13 on node testing-worker-linux-docker-bc362139-3367-linux-14.prod.travis-ci.org
```

Example: https://travis-ci.org/voxpupuli/puppet-archive/jobs/109740076

To fix this we can use getvar. This PR fixes that.

Thanks
Joseph

jyaworski commented:
I would rather use versioncmp and puppetversion as it's more idiomatic. We can sidestep aio_agent_version.

aerostitch commented:
Ok. I just saw #109, so I integrated the part that checks the version from @logicminds and also dropped the aio_agent_version fact call.

But it still needs to have the archive_windir fact checked for undef.

aerostitch commented:
Note that we don't take in account the specific case Antoine explained in https://tickets.puppetlabs.com/browse/FACT-1280 (having puppet running with 4.1 and system/os ruby when installed using the OS package on Fedora).
But we were not taking it in account in the first place either, and I'm not sure if it is possible at this stage.

logicminds commented:
To make things more complex the gem provider for windows was broken up until 4.3.1. I back ported a fix and made the pe_gem provider work on windows. So if your using the latest pe_gem provider code we can use the following code now.  This would require the usage of the latest pe_gem.

Note: this is untested, I think it would work

```puppet
if $::puppetversion =~ /Puppet Enterprise/  {
   # this is for puppet enterprise installs only, and only works on PE 3.x
    $gem_provider = 'pe_gem'  # now works on windows
} else {  
    if versioncmp($::puppetversion, '4.0.0') >= 0 {
      $gem_provider = 'puppet_gem'
    } else {
      $gem_provider = 'gem'
    }
}
```

aerostitch commented:
Thanks, in this case, let me just add a specific case for windows with puppet 4.3.1! ;)

logicminds commented:
Well technically you could have a special case for windows to always use pe_gem (even if not using pe) and then use puppet_gem if later than 4.3.0.  Although this might confuse people.

aerostitch commented:
Yep, that's the change I just pushed while you were hitting "comment"! ;)

aerostitch commented:
Updated the commit taking your advice in account. :)
Thanks

jyaworski commented on manifests/params.pp:
```diff
@@ -1,6 +1,19 @@
+    'Windows': {
+      $path               = getvar('::archive_windir')
+      $owner              = 'S-1-5-32-544' # Adminstrators
+      $group              = 'S-1-5-18'     # SYSTEM
```
What is this sorcery?

aerostitch replied:
I don't follow. What sorcery?

jyaworski replied:
Sorry, I was being facetious and referring to that S-1-5-32-544 'sorcery.'

aerostitch replied:
Looool!
That's the "Administrators" users according to the Windows convention: https://support.microsoft.com/en-us/kb/243330
No wonder why it's been years I haven't touched a Windows box! :)

logicminds commented on manifests/params.pp:
```diff
@@ -8,22 +21,11 @@
-  if $::puppetversion =~ /Puppet Enterprise/ and $::osfamily != 'Windows' {
-    $gem_provider = 'pe_gem'
-  } elsif $::aio_agent_version {
-    $gem_provider = 'puppet_gem'
```
The puppet_gem should be the default for non-windows and puppet > 4.0

aerostitch replied:
It is, check the lines added (green section) :)

In the default case:
```
 if $::puppetversion =~ /Puppet Enterprise/ or versioncmp($::puppetversion, '4.0.0') >= 0{
```

aerostitch commented:
Don't merge yet, I still need to do an update to get rid of the getvar usage.

jyaworski commented:
Please add [WIP] to the PR title.

jyaworski commented:
I just ran into this on [puppet-rundeck](https://travis-ci.org/voxpupuli/puppet-rundeck/jobs/110242268).

aerostitch commented:
Hi guys,

I updated the specs and params to fix the issue, cover the tests of different providers with newer versions of puppet and avoid using the getvar method for the archive_windir.
I also added a rspec test for the archive_windir fact at the same time.

Can you review and tell me if those changes are ok for you please?

Thanks
Joseph

aerostitch commented:
Note: I just added a note in the readme to help people that write modules depending on this module to avoid searching an issue for nothing when it comes to errors with the required fact puppetversion (not present in puppet-rspec by default). I think it could save some time to anyone integrating this module as a dependendcy of another one.

jyaworski commented on README.md:
````diff
@@ -87,6 +87,17 @@ class { 'archive':
+
+```
+RSpec.configure do |c|
+  c.default_facts = { :puppetversion => `puppet --version` }
````
You could use Puppet.version here.

jyaworski commented:
Besides my one comment on the README, this LGTM.

aerostitch commented:
Excellent advise on the Puppet.version.
Updated the PR accordingly.
Thanks! :)
